text/0000-delete-prefix.md

```diff
@@ -0,0 +1,135 @@
+belong to a member can become more efficient with delete by prefix.
+
+Since addresses are stored in a radix tree, deleting many addresses by one 
+prefix is much faster than deleting the same amount of addresses one by one.
+Furthermore, delete by prefix doesn't require transaction processor to know 
+which addresses exist. This will remove the need to store huge list of 
+existing addresses in a special address that will need to be modified by every
+transaction that writes a new address.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+A context API to delete address already exist for transaction processors.
+The existing delete API accepts only full 35 bytes addresses.
+This RFC exposes to transaction processor a new capability of deleting by 
+address prefix. When delete by prefix is called, all addresses under the 
+prefix subtree will be deleted.
+
+Main usage for this feature is the need to support some 'clean transaction'
+that will remove all addresses with some communality, for example all 
+addresses that belong to some member.
+The proposed solution is to have address mapping logic that allocates some
+unique pre-defined address prefix per member. The transaction processor logic
+will know to calculate the address suffix based on transaction payload data
+and execute any transaction.
+In order to execute the clean transaction, the transaction processor will
+call context.deletePrefix with the member prefix.
+
+The existing solution is to keep track of all used addresses, this will
+require read-modify-write of a special address for every new address that is
+used. This solution could work but will slow down executing of every regular 
+transactions while the clean transaction is rare.
+
+No change is recommended to the existing methods, so the implementation of this
+RFC should not cause any breakage for pre-existing transaction processors.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Add new API to context: deletePrefix that delete a subtree of addresses 
+starting from a given address prefix. 
+This API should be similar to regular delete address API but can get as input
+prefix of address and not just full 35 bytes address.
+Calling this API with full address should behave exactly like calling existing
+delete API.
+
+In the suggested implementation there are two proto changes:
+
+Addition to state_context.proto:
+	
+	// A request from the handler/tp to delete state prefix entries at an 
+	// collection of addresses prefix subtree
+	message TpStateDeletePrefixRequest {
+		string context_id = 1;
+		repeated string addresses_prefix = 2;
+	}
+
+	// A response form the contextmanager/validator with the addresses that 
+	// were deleted
+	message TpStateDeleteResponse {
+		enum Status {
+			STATUS_UNSET = 0;
+			OK = 1;
+			AUTHORIZATION_ERROR = 2;
+		}
+
+		repeated string addresses = 1;
```

Comment:
I'm not really sure if it it's necessary (or desirable) to return the list of addresses. It has the same performance problem of the previously suggested RFC for getting all the set addresses out of state.  These addresses would need to be read under that prefix, and then the prefix can be deleted. 

Deleting the prefix is cheap. Knowning what is under the prefix is not.

Reply:
agree, removed the list